pandas/core/dtypes/concat.py

```diff
@@ -58,6 +64,40 @@ def get_dtype_kinds(l):
     return typs
 
 
+def _cast_to_common_type(arr: ArrayLike, dtype: DtypeObj) -> ArrayLike:
+    """
+    Helper function for `arr.astype(common_dtype)` but handling all special
+    cases.
+    """
```

Comment:
do you have a plan in mind to make this function unnecessary?  it is special-casing pandas-internal EAs which i think we want to avoid

(Not a blocker for this PR because AFAICT this is necessary and we do this in the status quo anyway, just curious)

Reply:
Yeah, so this is indeed only needed to preserve value-based specific behaviour, not because I like this :)

Ideally we get rid of this, but I am not sure it is entirely possible with the numpy integer dtypes, since it will always depend on the presence of NaNs whether it can preserve int or not. The way to go here is probably the nullable integer dtypes to have consistent behaviour.